ir/visitor.cpp

```diff
@@ -20,76 +20,68 @@ limitations under the License.
 
 class Visitor::ChangeTracker {
     struct visit_info_t {
-        bool visit_finished;
+        bool visit_in_progress;
         const IR::Node *result;
     };
-    // typedef unordered_map<const IR::Node *, std::pair<bool, const IR::Node *>>  visited_t;
     typedef unordered_map<const IR::Node *, visit_info_t>  visited_t;
     visited_t           visited;
 
  public:
-    struct change_t {
-        // std::pair<visited_t::iterator, bool>    state;  // result of visited.emplace
-        bool           valid;
-        bool           already_present;
-        const IR::Node *original_node;
-        visit_info_t   *visit_info;    // Points into ChangeTracker.visited.
-
-        change_t() : valid(false) {}
-        change_t(visited_t *visited, const IR::Node *n) : valid(true) {
-                // Initialization
-                visited_t::iterator visited_it;
-                bool inserted;
-                std::tie(visited_it, inserted) =
-                    visited->emplace(n, visit_info_t({false,n}));
-                already_present = !inserted;
-                original_node = visited_it->first;
-                visit_info = &(visited_it->second);
-
-                // Sanity check for IR loops
-                if (!inserted && !visit_info->visit_finished)
-                    BUG("IR loop detected "); }
-
-        explicit operator bool() { return valid; }
-        bool done() { return already_present; }
-        const IR::Node *orig() { return original_node; }
-        const IR::Node *result() { return visit_info->result; }
-    };
 
-    bool done(const IR::Node *n) const {
-        return visited.count(n) && visited.at(n).visit_finished;
+    /* Begin tracking `n` during a visiting pass.  Use `finish` to mark `n` as
+     * visited once the pass completes.
+    */
+    void start(const IR::Node *n) {
+        // Initialization
+        visited_t::iterator visited_it;
+        bool inserted;
+        std::tie(visited_it, inserted) =
+            visited.emplace(n, visit_info_t({true,n}));
+
+        // Sanity check for IR loops
+        bool already_present = !inserted;
+        visit_info_t *visit_info = &(visited_it->second);
+        if (already_present && visit_info->visit_in_progress)
+            BUG("IR loop detected "); 
     }
-    const IR::Node *result(IR::Node *n) const { return visited.at(n).result; }
 
-    change_t track(const IR::Node *n) { return change_t(&visited, n); }
-    void start(change_t &change) { change.visit_info->visit_finished = false; }
-    bool finish(change_t &change, const IR::Node *orig, const IR::Node *final) {
+    /* Mark `n` as finished.  `done(n)` will return true, and `result(n)` will
```

Comment:
This makes sense.  I hadn't used Doxygen comments because this is an internal class to visitor.cpp, and I wasn't sure it made sense to expose that into the generated Doxygen documentation.  But the style you suggest is much clearer.